### ISSUE TITLE
"module is marked as livepatch module, but livepatch support is disabled"

### ISSUE THREAD

**Issue by jpoimboe** [dynup/kpatch#660]
When building a kpatch.ko based module for an upstream kernel with CONFIG_LIVEPATCH=n, it fails to load with:
```
[ 3817.975103] kpatch_kmalloc: module is marked as livepatch module, but livepatch support is disabled
```
I think the problem is that with #654, kpatch now sets MODULE_INFO(livepatch, "Y"), even for kpatch.ko based modules.  But the upstream kernel refuses to load any such modules if that modinfo flag is set.

Maybe we should remove that flag from kpatch-patch-hook.c and instead go back to using `add_taint()`.  To emulate add_taint_module() we could also do `set_bit(TAINT_LIVEPATCH, &mod->taints)`.

@joe-lawrence , thoughts?

**Comment by joe-lawrence** [comment]
argg, the case that #654 didn't test :frowning_face:

I think dropping ```MODULE_INFO(livepatch, "Y")``` and bringing back ```add_taint``` should get the module loaded.  ```set_bit(TAINT_LIVEPATCH)``` would then apply the per-module taint flag.  

Adjusting the taints flag directly feels a bit underhanded, though I guess not terrible since we would only be setting them.  Unfortunately I think this technique might require a RHEL-specific check since its notion of ```taints``` will be an unsigned int.

**Comment by jpoimboe** [comment]
> Adjusting the taints flag directly feels a bit underhanded, though I guess not terrible since we would only be setting them.

Agreed, though "underhanded" is kpatch's middle name :-)

> Unfortunately I think this technique might require a RHEL-specific check since its notion of taints will be an unsigned int.

Yes, though if we're going with this technique, we might consider reverting those patches since I guess they aren't needed after all?

**Comment by joe-lawrence** [comment]
> Yes, though if we're going with this technique, we might consider reverting those patches since I guess they aren't needed after all?

Good point.

What do you think about this alternative approach:
* In kpatch-patch-hook.c, wrap the MODULE_INFO(livepatch, "Y") around #ifdef RHEL >= 7.4

That would avoid us from having to fiddle with the taint flags from the module itself.

**Comment by jpoimboe** [comment]
> What do you think about this alternative approach:
> 
>     In kpatch-patch-hook.c, wrap the MODULE_INFO(livepatch, "Y") around #ifdef RHEL >= 7.4
> 
> That would avoid us from having to fiddle with the taint flags from the module itself.

But then what would be the taint strategy for other kernels?  Call set_taint()?

**Comment by joe-lawrence** [comment]
Right.  It's after 5pm and I'm not thinking straight :D

**Comment by joe-lawrence** [comment]
Fixed by PR #662, but @joe-lawrence botched the "Fixes" notation in the commit message, so it wasn't auto-closed.
